src/daft-plan/src/ops/project.rs

```diff
@@ -22,20 +25,276 @@ impl Project {
+                .values()
+                .cloned()
+                .collect::<Vec<_>>();
+            let plan: LogicalPlan =
+                Self::try_new(input, child_projection, resource_request.clone())?.into();
+            plan.into()
+        };
+        Ok((input, projection))
+    }
+
+    fn factor_expressions(exprs: &[Expr], schema: &Schema) -> (Vec<Expr>, IndexMap<String, Expr>) {
+        // Returns
+        //  1. original expressions with substitutions
+        //  2. substitution definitions
+        //
+        // E.g. factor_expressions([a+b as e, (a+b)+c as f]) ->
+        //  1. [some_id as e, some_id + c as f]
+        //  2. {some_id: a+b}
+
+        // Find all top-level repeated subexpressions;
+        // these are the expressions we will cache.
+        let mut seen_subexpressions = IndexSet::new();
+        let mut subexpressions_to_cache = IndexMap::new();
+
+        let mut exprs_to_walk: Vec<Arc<Expr>> = exprs.iter().map(|e| e.clone().into()).collect();
+        while !exprs_to_walk.is_empty() {
+            exprs_to_walk = exprs_to_walk
+                .iter()
+                .flat_map(|expr| {
+                    // If expr is an alias, ignore and continue recursing
+                    // (otherwise the unaliased child will be double counted)
+                    if matches!(expr.as_ref(), Expr::Alias(..)) {
+                        expr.children()
+                    } else {
+                        let expr_id = expr.semantic_id(schema);
+                        // Mark expr as seen
+                        let newly_seen = seen_subexpressions.insert(expr_id.clone());
+                        if newly_seen {
+                            // If not previously seen, continue recursing down children
+                            expr.children()
+                        } else {
+                            // If previously seen, cache the expression (if it involves computation)
+                            if optimization::requires_computation(expr) {
+                                subexpressions_to_cache.insert(expr_id.clone(), expr.clone());
+                            }
+                            // Stop recursing if previously seen;
+                            // we only want top-level repeated subexpressions
+                            vec![]
+                        }
+                    }
+                })
+                .collect();
+        }
+
+        // Then, substitute all the cached subexpressions in the original expressions.
+        let subexprs_to_replace = subexpressions_to_cache
+            .keys()
+            .cloned()
+            .collect::<IndexSet<_>>();
+        let substituted_expressions = exprs
+            .iter()
+            .map(|e| {
+                replace_column_with_semantic_id(e, &subexprs_to_replace, schema)
+                    .unwrap()
+                    .as_ref()
+                    .clone()
+            })
+            .collect::<Vec<_>>();
```

Comment:
If we change `replace_column_with_semantic_id` to have move semantics and change `exprs` to an owned `Vec<Expr>`, we shouldn't need the clone:
```suggestion
            .into_iter()
            .map(|e| {
                replace_column_with_semantic_id(e, &subexprs_to_replace, schema)
                    .unwrap()
            })
            .collect::<Vec<_>>();
```